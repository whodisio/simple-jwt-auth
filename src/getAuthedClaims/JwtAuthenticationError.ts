import { SimpleJwtAuthError } from '../SimpleJwtAuthError';

export class JwtAuthenticationError extends SimpleJwtAuthError {
  constructor({ reason }: { reason: string }) {
    const message = `
this JWT can not be trusted! ${reason}
    `.trim();
    super(message);
  }
}
