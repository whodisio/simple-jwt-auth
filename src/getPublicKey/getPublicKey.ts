import { SimpleJwtAuthError } from '../SimpleJwtAuthError';
import { getUnauthedClaims } from '../getUnauthedClaims';
import { getUnauthedHeaderClaims } from '../getUnauthedHeaderClaims';
import { discoverJwksUriFromAuthServerMetadata } from './discoverJwksUriFromAuthServerMetadata';
import { extractPublicKeyFromJwksUri } from './extractPublicKeyFromJwksUri';
import { cachePublicKey, getPublicKeyFromCache } from './publicKeyCache';

export class GetPublicKeyOfTokenError extends SimpleJwtAuthError {
  constructor({ reason }: { reason: string }) {
    const message = `
${reason}

Can not get public key of token
    `.trim();
    super(message);
  }
}

/**
 * gets the public key used to sign the token
 *
 * note
 * - caches the results to prevent redundant network requests
 * - supports oauth discovery flow, to discover the jwks endpoint from the issuer
 * - supports manual flow, to explicitly define the jwks endpoint for issuers who dont support discovery
 */
export const getPublicKey = async ({
  token,
  jwksUri: jwksUriInput,
  cache = { ttlInSeconds: 300 },
}: {
  token: string;
  jwksUri?: string;
  cache?: { ttlInSeconds: number };
}) => {
  // grab the issuer from the token
  const claims = getUnauthedClaims({ token });
  const issuer = claims.iss;
  if (!issuer)
    throw new GetPublicKeyOfTokenError({
      reason: 'Issuer not defined on the token (i.e., no `claims.iss`).',
    });

  // grab the keyId from the token
  const headerClaims = getUnauthedHeaderClaims({ token });
  const keyId = headerClaims.kid;
  if (!keyId)
    throw new GetPublicKeyOfTokenError({
      reason: 'KeyId is not defined on the token (i.e., no `header.kid`)',
    });

  // try to return the public key if we already have it in the cache
  const cachedKey = getPublicKeyFromCache({ issuer, keyId });
  if (cachedKey) return cachedKey; // if we got it, we got it

  // otherwise, lookup the public key
  const jwksUri =
    jwksUriInput ?? (await discoverJwksUriFromAuthServerMetadata({ claims }));
  const publicKey = await extractPublicKeyFromJwksUri({
    headerClaims,
    jwksUri,
  });

  // cache the public key for future lookups (to save on the two HTTP calls)
  cachePublicKey({
    issuer,
    keyId,
    publicKey,
    ttlInSeconds: cache.ttlInSeconds,
  });

  // return the pem cert
  return publicKey; // this is the public key
};
