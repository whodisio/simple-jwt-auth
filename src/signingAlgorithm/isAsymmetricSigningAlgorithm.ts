const ASYMMETRIC_SIGNING_ALGORITHMS = ['RS256', 'RS384', 'RS512'] as const;
export type AsymmetricSigningAlgorithm =
  typeof ASYMMETRIC_SIGNING_ALGORITHMS[number];
export const isAsymmetricSigningAlgorithm = (
  alg: any,
): alg is AsymmetricSigningAlgorithm =>
  ASYMMETRIC_SIGNING_ALGORITHMS.includes(alg);
