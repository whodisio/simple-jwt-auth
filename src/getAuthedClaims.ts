import { SimpleJwtAuthError } from './SimpleJwtAuthError';
import { getPublicKey } from './getPublicKey/getPublicKey';
import { getSignedClaims } from './getSignedClaims';
import { MinimalTokenClaims } from './getUnauthedClaims';
import { verifyTokenIntent } from './verification/verifyTokenIntent';
import { verifyTokenShape } from './verification/verifyTokenShape';
import { verifyTokenTimestamps } from './verification/verifyTokenTimestamps';

/**
 * Authenticates the claims made by the JWT, conforming to highest security standards, before returning claims.
 *
 * Checks:
 * - that the token was issued by the expected issuer (prevents trusting malicious issuers)
 * - that the token was issued to be used by the expected audience (prevents trusting tokens issued for a different application to be hijacked used in yours)
 * - that the token is still valid based on timestamps (prevents having tokens that live forever - which are a greater risk if stolen)
 * - that an asymmetric signing algorithm was used (prevents accidental usage of symmetric signing algorithms in distributed setting)
 * - that the issuer really made these claims, by checking the signature (prevents malicious users from manipulating the claims in a token after its been issued)
 */
export const getAuthedClaims = async <C extends MinimalTokenClaims>({
  token,
  issuer,
  audience,
  jwksUri,
}: {
  token: string;
  issuer: string;
  audience: string | string[];

  /**
   * the jwks uri at which the public key for this token can be found
   *
   * note
   * - for issuer's who support oauth2's discovery flow, this can be left blank, and the uri will be discovered (recommended)
   * - for issuer's who unable to support oauth2's discovery flow, this will need to be explicitly defined
   */
  jwksUri?: string;
}): Promise<C> => {
  // runtime validation: confirm and audiences were defined (everyone has types until they get punched in the runtime - mike tyson)
  if (!issuer)
    throw new SimpleJwtAuthError(
      'expected issuer must be defined for secure distributed jwt authentication',
    );
  if (!audience)
    throw new SimpleJwtAuthError(
      'expected audience must be defined for secure distributed jwt authentication',
    );
  const audiences = Array.isArray(audience) ? audience : [audience]; // normalize it to a more generic form in the process
  if (!audiences.length)
    throw new SimpleJwtAuthError(
      'at least one expected audience must be defined for secure distributed jwt authentication',
    );

  // check that the token has standard and expected shape
  await verifyTokenShape({ token });

  // check whether the audience and issuer of the token match, before even looking at verifying signature; critical for security in distributed system
  await verifyTokenIntent({
    token,
    intendedIssuer: issuer,
    intendedAudiences: audiences,
  });

  // check the timestamps to confirm the token is not expired
  await verifyTokenTimestamps({ token });

  // check that the token was signed correctly and safely; if so, grab claims
  const publicKey = await getPublicKey({ token, jwksUri });
  const signedClaims = await getSignedClaims<C>({ token, publicKey });

  // since the claims passed all of the checks above, they have now been authenticated
  return signedClaims;
};
