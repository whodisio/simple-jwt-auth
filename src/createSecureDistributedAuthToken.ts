import { base64UrlEncode } from './base64Url/base64UrlEncode';
import { MinimalTokenClaims } from './getUnauthedClaims';
import { MinimalTokenHeaderClaims } from './getUnauthedHeaderClaims';
import { isAsymmetricSigningAlgorithm } from './signingAlgorithm/isAsymmetricSigningAlgorithm';
import { SimpleJwtAuthError } from './SimpleJwtAuthError';
import { createVerifiableSignature } from './signingAlgorithm/createVerifiableSignature';

/**
 * Creates secure, authenticatable tokens for a distributed system - enforcing security standards
 */
export const createSecureDistributedAuthToken = <C extends MinimalTokenClaims>({
  headerClaims,
  claims,
  privateKey,
}: {
  headerClaims: MinimalTokenHeaderClaims;
  claims: C;
  privateKey: string;
}): string => {
  // check that the signing algorithm is asymmetric
  if (!isAsymmetricSigningAlgorithm(headerClaims.alg))
    throw new SimpleJwtAuthError('only asymmetric signing algorithms are allowed in distributed systems');

  // check that an issuer and audience have been defined
  if (!claims.iss) throw new SimpleJwtAuthError('token.claims.iss must be defined when creating a secure token');
  if (!claims.aud) throw new SimpleJwtAuthError('token.claims.aud must be defined when creating a secure token');

  // check that expiration has been defined
  if (!claims.exp) throw new SimpleJwtAuthError('token.claims.exp must be defined when creating a secure token');

  // create a token
  const payload = [base64UrlEncode(JSON.stringify(headerClaims)), base64UrlEncode(JSON.stringify(claims))].join('.');
  const signature = createVerifiableSignature({ alg: headerClaims.alg, payload, privateKey });
  const token = [...payload.split('.'), signature].join('.');

  // return the token
  return token;
};
