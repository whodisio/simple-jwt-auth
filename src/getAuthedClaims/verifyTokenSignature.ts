import crypto from 'crypto';

import { discoverPublicKeyFromAuthServerMetadata } from '../discoverPublicKeyFromAuthServerMetadata/discoverPublicKeyFromAuthServerMetadata';
import { getUnauthedHeaderClaims } from '../getUnauthedHeaderClaims';
import { JwtAuthenticationError } from './JwtAuthenticationError';
import { castJwtAlgToCryptoAlg } from '../signingAlgorithm/castJwtAlgToCryptoAlg';

const castBase64UrlToBase64 = (base64Url: string) => base64Url.replace(/-/g, '+').replace(/_/g, '/');

export const verifyTokenSignature = async ({ token }: { token: string }) => {
  // grab the public key in a distributed fashion, w/ standard oauth discovery flow
  const publicKey = await discoverPublicKeyFromAuthServerMetadata({ token });

  // convert the alg the token said it was signed by to an alg name that crypto understands
  const unauthedHeaderClaims = getUnauthedHeaderClaims({ token });
  const cryptoAlg = castJwtAlgToCryptoAlg(unauthedHeaderClaims.alg);

  // check that we can verify the signature
  const verificationInput = token.split('.').slice(0, 2).join('.'); // drop the signature
  const signatureClaim = token.split('.').slice(-1)[0]; // signature is the last el
  const unescapedSignatureClaim = castBase64UrlToBase64(signatureClaim); // JWT's are url encoded, so that they are safe to pass in urls (i.e., base64URL format). crypto only supports normal base64, so we decode base64Url into base64 for interop
  const verified = crypto.createVerify(cryptoAlg).update(verificationInput).verify(publicKey, unescapedSignatureClaim, 'base64');
  if (!verified) throw new JwtAuthenticationError({ reason: 'signature is incorrect' });
};
