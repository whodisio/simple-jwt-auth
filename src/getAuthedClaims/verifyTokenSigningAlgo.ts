import { getUnauthedHeaderClaims } from '../getUnauthedHeaderClaims';
import { isAsymmetricSigningAlgorithm } from '../signingAlgorithm/isAsymmetricSigningAlgorithm';
import { JwtAuthenticationError } from './JwtAuthenticationError';

/**
 * only asymmetric signing algorithms can be securely used in a distributed authentication strategy.
 *
 * why? symmetric signing algorithms do not have public keys, so third parties have no way of verifying the authenticity of a token - unless the private key is made public, in which case anyone can issue verifiable claims and the strategy is no longer secure.
 *
 * this library supports distributed authentication strategies, so symmetric keys are forbidden
 */
export const verifyTokenSigningAlgo = ({ token }: { token: string }) => {
  const unauthedHeaderClaims = getUnauthedHeaderClaims({ token });
  if (!isAsymmetricSigningAlgorithm(unauthedHeaderClaims.alg))
    throw new JwtAuthenticationError({
      reason: `tokens must be signed with asymmetric signing algorithms for secure distributed jwt authentication. found \`token.header.alg = ${unauthedHeaderClaims.alg}\` instead.`,
    });
};
