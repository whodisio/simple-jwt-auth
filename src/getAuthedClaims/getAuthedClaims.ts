import { MinimalTokenClaims, getUnauthedClaims } from '../getUnauthedClaims';
import { SimpleJwtAuthError } from '../SimpleJwtAuthError';
import { verifyTokenIntent } from './verifyTokenIntent';
import { verifyTokenSigningAlgo } from './verifyTokenSigningAlgo';
import { verifyTokenTimestamps } from './verifyTokenTimestamps';
import { verifyTokenSignature } from './verifyTokenSignature';

/**
 * Authenticates the claims made by the JWT, conforming to highest security standards, before returning claims.
 *
 * This method is primarily used in server side / backend applications, where we must trust the data we are working with.
 *
 * In client side / frontend applications, there is no reason to authenticate claims as no data on frontend should be trusted anyway: use `getUnauthedClaims` instead.
 */
export const getAuthedClaims = async <C extends MinimalTokenClaims>({
  token,
  issuer,
  audience,
}: {
  token: string;
  issuer: string;
  audience: string | string[];
}): Promise<C> => {
  // runtime validation: confirm and audiences were defined (everyone has types until they get punched in the runtime - mike tyson)
  if (!issuer) throw new SimpleJwtAuthError('expected issuer must be defined for secure distributed jwt authentication');
  if (!audience) throw new SimpleJwtAuthError('expected audience must be defined for secure distributed jwt authentication');
  const audiences = Array.isArray(audience) ? audience : [audience]; // normalize it to a more generic form in the process
  if (!audiences.length) throw new SimpleJwtAuthError('at least one expected audience must be defined for secure distributed jwt authentication');

  // check whether the audience and issuer of the token match, before even looking at verifying signature; critical for security in distributed system
  await verifyTokenIntent({ token, intendedIssuer: issuer, intendedAudiences: audiences });

  // check the timestamps to confirm the token is not expired
  await verifyTokenTimestamps({ token });

  // check that the token was signed by an asymmetric algorithm - otherwise, this token can't be used securely in a distributed system. (i.e., there is no "publicKey" in a symmetric signing algo)
  await verifyTokenSigningAlgo({ token });

  // verify that the token was signed by the issuer
  await verifyTokenSignature({ token });

  // return the claims of the token, as with the above checks we have authenticated those claims
  const tokenClaims = getUnauthedClaims<C>({ token });
  return tokenClaims; // they are now authed
};
