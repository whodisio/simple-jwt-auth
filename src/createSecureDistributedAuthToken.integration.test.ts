import { given } from 'test-fns';

import { createSecureDistributedAuthToken } from './createSecureDistributedAuthToken';
import { getAuthedClaims } from './getAuthedClaims';
import { getPublicKey } from './getPublicKey/getPublicKey';
import {
  exampleEccKeyPair,
  exampleRsaKeyPair,
} from './signingAlgorithm/createVerifiableSignature.test';
import { AsymmetricSigningAlgorithm } from './signingAlgorithm/isAsymmetricSigningAlgorithm';

// mock that we discover the real public key each time
jest.mock('./getPublicKey/getPublicKey');
const getPublicKeyMock = getPublicKey as jest.Mock;

describe('createSecureToken', () => {
  given('rsa signing algorithm', () => {
    const alg: AsymmetricSigningAlgorithm = 'RS256';
    const keypair = exampleRsaKeyPair;

    beforeEach(() => {
      getPublicKeyMock.mockReturnValue(keypair.publicKey);
    });

    it('should be able to create a token that we can later getAuthedClaims on', async () => {
      // create a token
      const token = createSecureDistributedAuthToken({
        headerClaims: { alg, kid: '4.some_directory', typ: 'JWT' },
        claims: {
          iss: 'https://auth.whodis.io/...',
          aud: '__some_directory__',
          sub: '__some_user__',
          exp: 2516239022,
        },
        privateKey: keypair.privateKey,
      });
      expect(typeof token).toEqual('string'); // sanity check

      // check that we can auth on it, if the publicKey is discoverable
      const claims = await getAuthedClaims({
        token,
        issuer: 'https://auth.whodis.io/...',
        audience: '__some_directory__',
      });
      expect(claims).toEqual({
        iss: 'https://auth.whodis.io/...',
        aud: '__some_directory__',
        sub: '__some_user__',
        exp: 2516239022,
      });
    });
  });

  given('ecc signing algorithm', () => {
    const alg: AsymmetricSigningAlgorithm = 'ES256';
    const keypair = exampleEccKeyPair;

    beforeEach(() => {
      getPublicKeyMock.mockReturnValue(keypair.publicKey);
    });

    it('should be able to create a token that we can later getAuthedClaims on', async () => {
      // create a token
      const token = createSecureDistributedAuthToken({
        headerClaims: { alg, kid: '4.some_directory', typ: 'JWT' },
        claims: {
          iss: 'https://auth.whodis.io/...',
          aud: '__some_directory__',
          sub: '__some_user__',
          exp: 2516239022,
        },
        privateKey: keypair.privateKey,
      });
      expect(typeof token).toEqual('string'); // sanity check

      // check that we can auth on it, if the publicKey is discoverable
      const claims = await getAuthedClaims({
        token,
        issuer: 'https://auth.whodis.io/...',
        audience: '__some_directory__',
      });
      expect(claims).toEqual({
        iss: 'https://auth.whodis.io/...',
        aud: '__some_directory__',
        sub: '__some_user__',
        exp: 2516239022,
      });
    });
  });
});
