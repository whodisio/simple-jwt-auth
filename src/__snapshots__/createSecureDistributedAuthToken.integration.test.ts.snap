// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createVerifiableSignature given: rsa algorithm when: provided an rsa key then: create verifiable signature 1`] = `"GTpV_xVR7RBtkI9IFvi9ct6nixEas80j8DSuE0NRWcPjZk02opg-Ifc980HWuaBr-Kcn5-8jtSFKprl4g_Z_n4AgMQInYn63gMNneEfmeVOeqTBvJ4I03nSQ3StKydZk1G5rDm42FTjA6QfVDwojWDP19RC1eLhlUcAhQCr_mh7F23Qul8_B9c6pO1UrMKvYv9LT4USuffDY7E0IDj6HS97CrjYJIrmYeUkQFbWhJxzC6euTaRyqDiQfsSpc0MYcN-mwFwUVMtKkhlfnAq5IC68iqSP__jvpEhSFo8PkuSLFc3-1FLdBSYb6KqSvx6gtca60P16681RLkhRQKx5nCEGcDVexm1W_totb9ZcFxCHp19BJU7XgdCf2uLJCmruIMIwJ5uQhcU9axTShR7005tqDLm0wn_pLG1ckKtsOzcmtR8li34HPdfN39CWL0nmPGIgLQKzBfvnqzVYS9-8g6ddqnfDzBOgia-TyNC599ZLK2-8yrtb3kbclSBRDM2yD3CFbHBdZ9q-sLdbnQ8zohhd1R5pNq0JutF7b6Pm-SrjDkMxKbG5eIA8-4J0fh1rqScjQzoS5Kv_OjQwQ2dZKubGkR9CFuDnPLbkkbjsB0J-5Bx-2AU2RyjW0_QID87lw9k18LaVQWvkkECHis-sPXkfwbRoJHa5b5ww-sT_pcJU"`;
